apisix/plugins/authz-casdoor.lua

```diff
@@ -0,0 +1,172 @@
+    },
+    required = {
+        "callback_url", "endpoint_addr", "client_id", "client_secret"
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 2559,
+    name = plugin_name,
+    schema = schema
+}
+
+local function fetch_access_token(ctx, conf, state_in_session)
+    local args = core.request.get_uri_args(ctx)
+    local client = http.new()
+    local url = conf.endpoint_addr .. "/api/login/oauth/access_token"
+
+    local res, err = client:request_uri(url, {
+        method = "POST",
+        body =  ngx.encode_args({
+            code = args.code,
+            grant_type = "authorization_code",
+            client_id = conf.client_id,
+            client_secret = conf.client_secret
+        }),
+        headers = {
+            ["Content-Type"] = "application/x-www-form-urlencoded"
+        }
+    })
+
+    if not res then
+        return nil, nil, err
+    end
+    local data, err = core.json.decode(res.body)
+
+    if err or not data then
+        err = "failed to parse casdoor response data: " .. err .. ", body: " .. res.body
+        return nil, nil, err
+    end
+
+    if not data.access_token then
+        return nil, nil,
+               "failed when accessing token: no access_token contained"
+    end
+    -- In the reply of casdoor, setting expires_in to 0 indicates that the access_token is invalid.
+    if not data.expires_in or data.expires_in == 0 then
+        return nil, nil, "failed when accessing token: invalid access_token"
+    end
+
+    return data.access_token, data.expires_in, nil
+end
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.access(conf, ctx)
+    local current_uri = ctx.var.uri
+    local session_obj_read, session_present = session.open()
+    -- step 1: check whether hits the callback
+    local m, err = ngx.re.match(conf.callback_url, ".+//[^/]+(/.*)", "jo")
+    if err or not m then
+        core.log.error(err)
+        return 503
+    end
+    local real_callback_url = m[1]
+    if current_uri == real_callback_url then
+        if not session_present then
+            err = "no session found"
+            core.log.error(err)
+            return 503
+        end
+        local state_in_session = session_obj_read.data.state
+        if not state_in_session then
+            err = "no state found in session"
+            core.log.error(err)
+            return 503
+        end
+        local args = core.request.get_uri_args(ctx)
+        if not args or not args.code or not args.state then
+            err = "failed when accessing token. Invalid code or state"
+            core.log.error(err)
+            return 400, err
+        end
+        if args.state ~= tostring(state_in_session) then
+            err = "invalid state"
+            core.log.error(err)
+            return 400, err
+        end
+        local access_token, lifetime, err =
+            fetch_access_token(ctx, conf, state_in_session)
```

Comment:
We can pass args instead of ctx to this function. So that the scope is limited.

Reply:
Done